### PR TITLE
Add Media datasource/import and enhanced read

### PR DESCRIPTION
**Pull request by vbauzysvmware**
* Add Media datasource
* Add Media import
* Add a few `computed` fields to resource read
* Improved read - fixed issues
* Changed ID from constructed name to ID

### CODE REVIEW

File: vcd/catalogitem.go [hunk 1]
```diff
@@ -55,18 +56,30 @@ func findCatalogItem(d *schema.ResourceData, vcdClient *VCDClient) (*govcd.Catal
...
+
+	catalogItem, err := catalog.GetCatalogItemByNameOrId(identifier, false)
 	if err != nil {
 		log.Printf("[DEBUG] Unable to find catalog item. Removing from tfstate")
-		return nil, fmt.Errorf("unable to find catalog item: %s", err)
+		d.SetId("")
+		return nil, nil
+	}
+
+	entityId, err := govcd.GetBareEntityUuid(catalogItem.CatalogItem.ID)
```
Comment (dataclouder) [inline vcd/catalogitem.go]: Why we can't use the simple catalogItem.ID instead of doing this transformation?

Comment (vbauzysvmware) [reply vcd/catalogitem.go]: Cause anyways we will need todo transformation when we use or refernce ID and you have to know, that you need to parse it. So this chage allow to save "384be749-2d58-41e6-b4a0-0ce407748cea" and not "urn:vcloud:media:84be749-2d58-41e6-b4a0-0ce407748cea"

Comment (dataclouder) [reply vcd/catalogitem.go]: It seems to me that we are adding complexity to terraform-provider-vcd instead of dealing with it in go-vcloud-director.

I would expect to have methods that retrieve a media item with all the elements that are needed here, and then we simply use it.
One thing I don't understand is why sometimes we look for MediaItem through a Query and sometimes we use GetCatalogItem.

Comment (vbauzysvmware) [reply vcd/catalogitem.go]: This I think is wider discussion regarding what type of ID we save and fetch from state file. 
Regarding MediaItem and CatalogItem as Media -> you can upload Media to catalog and to vdc. Also mediaitem resides as catalogitem in catalog. Its a mess already in vCD.

File: CHANGELOG.md [hunk 2]
```diff
@@ -41,6 +42,8 @@ IMPROVEMENTS:
 * `vcd_org_vdc` Add import capability and full read support [#218]
 * Upgrade Terraform SDK dependency to 0.12.8 [#320]
 * `resource/vcd_vapp_vm` has new field `computer_name` [#334]
+* `resource/vcd_catalog_media` Add computed properties (is_iso, owner_name, is_published, creation_date, size, status, storage_profile_name) and full read support [#340]
```
Comment (lvirbalas) [inline CHANGELOG.md]: Please quote the property names.

File: CHANGELOG.md [hunk 3]
```diff
@@ -41,6 +42,8 @@ IMPROVEMENTS:
...
+* `resource/vcd_catalog_media`: MediaItem state ID changed from colon separated list of catalog name and media name to vCD ID [#340]
```
Comment (lvirbalas) [inline CHANGELOG.md]: The colon ":" is extra - please remove it here and from a few entries above in this section to make it look consistent.

File: vcd/catalogitem.go [hunk 4]
```diff
@@ -55,18 +56,30 @@ func findCatalogItem(d *schema.ResourceData, vcdClient *VCDClient) (*govcd.Catal
...
-		d.SetId("")
+		log.Printf("[DEBUG] Unable to find catalog.")
 		return nil, fmt.Errorf("unable to find catalog: %s", err)
 	}
 
-	catalogItem, err := catalog.GetCatalogItemByName(d.Get("name").(string), false)
+	identifier := d.Id()
+
+	// check if identifier still is in deprecated style `catalogName:mediaName`
+	if identifier == "" || strings.Contains(identifier, ":") {
```
Comment (lvirbalas) [inline vcd/catalogitem.go]: Is this needed for a smooth upgrade from 2.4.0 to 2.5.0?

File: vcd/resource_vcd_catalog_item.go [hunk 5]
```diff
@@ -187,23 +193,10 @@ func createOrUpdateCatalogItemMetadata(d *schema.ResourceData, meta interface{})
...
-		return fmt.Errorf(errorRetrievingOrg, err)
-	}
-
-	catalog, err := adminOrg.GetCatalogByName(d.Get("catalog").(string), false)
-	if err != nil {
-		log.Printf("[DEBUG] Unable to find catalog: %s", err)
-		return nil
-	}
-
-	catalogItem, err := catalog.GetCatalogItemByName(d.Get("name").(string), false)
```
Comment (lvirbalas) [inline vcd/resource_vcd_catalog_item.go]: Please tell more why this code became not needed after this PR?

File: website/docs/d/catalog_media.html.markdown [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+## Example Usage
+
+```hcl
+data "vcd_catalog_media" "existing-media" {
+  org     = "my-org"
+  catalog = "my-cat"
+  name    = "my-media"
+}
+
+output "gateway" {
```
Comment (lvirbalas) [inline website/docs/d/catalog_media.html.markdown]: gateway?

File: website/docs/d/catalog_media.html.markdown [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  value = data.vcd_catalog_media.size
+}
+
+output "dns1" {
```
Comment (lvirbalas) [inline website/docs/d/catalog_media.html.markdown]: dns1?

File: website/docs/r/catalog_media.html.markdown [hunk 8]
```diff
@@ -45,3 +45,45 @@ The following arguments are supported:
...
+
+## Importing
+
+Supported in provider *v2.5+*
+
+~> **Note:** The current implementation of Terraform import can only import resources into the state. It does not generate
+configuration. [More information.][docs-import]
+
+An existing media item can be [imported][docs-import] into this resource via supplying its path.
+The path for this resource is made of orgName.catalogName.mediaName.
```
Comment (lvirbalas) [inline website/docs/r/catalog_media.html.markdown]: I suggest to use notion with dashes everywhere. I.e. "org-name.catalog-name.media-name" - it's just more consistent and easier to scan with eyes.

File: vcd/catalogitem.go (hunk 1 above)
Comment (lvirbalas) [reply vcd/catalogitem.go]: > This I think is wider discussion regarding what type of ID we save and fetch from state file.

What's the question specifically?

> you can upload Media to catalog and to vdc

Please elaborate on how to upload to VDC only?

File: CHANGELOG.md (hunk 2 above)
Comment (vbauzysvmware) [reply CHANGELOG.md]: done

File: CHANGELOG.md (hunk 3 above)
Comment (vbauzysvmware) [reply CHANGELOG.md]: done

File: vcd/catalogitem.go (hunk 4 above)
Comment (vbauzysvmware) [reply vcd/catalogitem.go]: Yes, cause in state file we could get old style id

File: vcd/resource_vcd_catalog_item.go (hunk 5 above)
Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog_item.go]: Just removed repeatable code which was in already existing function

File: website/docs/d/catalog_media.html.markdown (hunk 6 above)
Comment (vbauzysvmware) [reply website/docs/d/catalog_media.html.markdown]: fixed

File: website/docs/d/catalog_media.html.markdown (hunk 7 above)
Comment (vbauzysvmware) [reply website/docs/d/catalog_media.html.markdown]: fixed

File: website/docs/r/catalog_media.html.markdown (hunk 8 above)
Comment (vbauzysvmware) [reply website/docs/r/catalog_media.html.markdown]: fixed

File: vcd/catalogitem.go (hunk 1 above)
Comment (vbauzysvmware) [reply vcd/catalogitem.go]: Question what type of Id save in state file:
urn:vcloud:media:84be749-2d58-41e6-b4a0-0ce407748cea
vs
384be749-2d58-41e6-b4a0-0ce407748cea
----
Seems we have to save in extended ID style and parse UUID when we need, as NSX API requests extended ID style.

Comment (vbauzysvmware) [reply vcd/catalogitem.go]: After analysis - API allows to upload to vCD or catalog. Our current govcd supports only to catalog way.

File: vcd/catalogitem.go (hunk 4 above)
Comment (lvirbalas) [reply vcd/catalogitem.go]: Got it. Please change/extend the comment with something like this:
```
// Check if identifier is still in deprecated style `catalogName:mediaName`
// Required for backwards compatibility as identifier has been changed to vCD ID in 2.5.0
```

File: website/docs/r/catalog_media.html.markdown (hunk 8 above)
Comment (lvirbalas) [reply website/docs/r/catalog_media.html.markdown]: Thanks, just one more ask: there's a dot at the end of the sentence, which may trip inexperienced user, please remove it as well.

File: vcd/catalogitem.go (hunk 1 above)
Comment (lvirbalas) [reply vcd/catalogitem.go]: You mean VDC? Interesting! Would you share a screen shot how to do it through UI?

Comment (lvirbalas) [reply vcd/catalogitem.go]: > Seems we have to save in extended ID style and parse UUID when we need, as NSX API requests extended ID style.

And what ID styles other resources use?

Comment (vbauzysvmware) [reply vcd/catalogitem.go]: Atleast I need short one cause usually we construct api request URL with short ID.

File: website/docs/r/catalog_media.html.markdown (hunk 8 above)
Comment (vbauzysvmware) [reply website/docs/r/catalog_media.html.markdown]: done

File: vcd/catalogitem.go (hunk 4 above)
Comment (vbauzysvmware) [reply vcd/catalogitem.go]: done

File: vcd/catalogitem.go (hunk 1 above)
Comment (vbauzysvmware) [reply vcd/catalogitem.go]: > You mean VDC? Interesting! Would you share a screen shot how to do it through UI?

From UI perspective - in flash I don't have working plugin - so haven't seen myself. In HTML5 UI - found only for Tenant upload capabilities, but both media and OVF requires provide catalog.

Comment (lvirbalas) [reply vcd/catalogitem.go]: Oh, so it's API-only capability?

Comment (vbauzysvmware) [reply vcd/catalogitem.go]: I know it's in API and in UI I can't confirm due reasons of flash current compatibility and in new HTML5 I didn't found.

Comment (lvirbalas) [reply vcd/catalogitem.go]: @dataclouder @Didainius we discussed the following three points in-depth with Vaidotas and converged to the following conclusions:

1. Use the `urn:...` long ID format, because we want to be a thin layer above vCD API where possible. Granted, this adds additional `GetBareEntityUuid` calls inside the code.
2. Continue using `Query...` functions where it makes sense. `Query...` functions are by agreement those, which call upon the `/query` API endpoint. Most of them return array, but some may return a filtered single record. We already have both return types in code, but it's no issue. In general, we had an agreement that we have `Get` and `Query` function classes some time ago. In this case it's easier to use /query, because it's only a single call to get the media, as opposed to getting it through the catalog.
3. Review which media functions may need to be deprecated. There's one `Find...` function that certainly has a bug - it errors out if there are two media items with same name in two catalogs. We want to move away from it.

Next step is for Vaidotas to refactor where appropriate given the above.

Comment (dataclouder) [reply vcd/catalogitem.go]: This decision goes against our own Coding guidelines. (https://github.com/vmware/go-vcloud-director/blob/master/CODING_GUIDELINES.md)

Comment (vbauzysvmware) [reply vcd/catalogitem.go]: I already added all Get* functions by code guidelines, Query is additional one to get more data.

File: vcd/resource_vcd_catalog_media.go [hunk 9]
```diff
@@ -152,19 +191,40 @@ func resourceVcdMediaRead(d *schema.ResourceData, meta interface{}) error {
...
+	if err != nil {
+		log.Printf("[DEBUG] Unable to find media item: %s", err)
+		return err
+	}
+	if catalogItem == nil {
+		log.Printf("[DEBUG] Unable to find media item: %s. Removing from tfstate", err)
+		return nil
+	}
+
+	mediaItem, err := vdc.QueryMediaImage(catalogItem.CatalogItem.Name, d.Get("catalog").(string))
```
Comment (dataclouder) [inline vcd/resource_vcd_catalog_media.go]: This call will fail. 
There is no VDC in the media arguments, and there is no guarantee that there will be such parameter in the provider section, since we can use catalogs and media without mentioning a VDC.
Please use `Catalog.GetMedia*` to retrieve the image

File: vcd/datasource_vcd_catalog_media.go [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+				Computed: true,
+			},
+			"metadata": {
+				Type:        schema.TypeMap,
+				Computed:    true,
+				Description: "Key and value pairs for catalog item metadata",
+				// For now underlying go-vcloud-director repo only supports
+				// a value of type String in this map.
+			},
+			"is_iso": &schema.Schema{
```
Comment (dataclouder) [inline vcd/datasource_vcd_catalog_media.go]: This question should have been asked when the resource was created, but let's see it here:
Why properties `is_iso` and `is_published` are strings instead of Boolean, and why is `size` a string instead of a number?

File: vcd/resource_vcd_catalog_media.go (hunk 9 above)
Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog_media.go]: it's not an option cause Catalog.GetMedia don't return all attributes I need

File: vcd/resource_vcd_catalog_media.go [hunk 11]
```diff
@@ -226,3 +286,62 @@ func createOrUpdateMediaItemMetadata(d *schema.ResourceData, meta interface{}) e
...
+	if err != nil {
+		return nil, fmt.Errorf(errorRetrievingOrg, orgName)
+	}
+
+	catalog, err := adminOrg.GetCatalogByName(catalogName, false)
+	if err != nil {
+		return nil, govcd.ErrorEntityNotFound
+	}
+
+	catalogItem, err := catalog.GetCatalogItemByName(catalogItemName, false)
```
Comment (dataclouder) [inline vcd/resource_vcd_catalog_media.go]: Why not using `GetMedia*`?

File: vcd/datasource_vcd_catalog_media.go (hunk 10 above)
Comment (vbauzysvmware) [reply vcd/datasource_vcd_catalog_media.go]: Cause they are computed - so no difference, but I look if there is other implications

File: vcd/resource_vcd_catalog_media.go (hunk 11 above)
Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog_media.go]: cause I haven't updated this PR yet

File: vcd/catalogitem.go (hunk 1 above)
Comment (lvirbalas) [reply vcd/catalogitem.go]: > This decision goes against our own Coding guidelines
It's not good that coding guidelines do not describe Query parameters. We had an agreement that for /query endpoints we keep `Query...` functions. I don't see how `Get...` functions can replace `Query...` functions, especially those which return an array. Please amend the coding guidelines or let's discuss.

Comment (dataclouder) [reply vcd/catalogitem.go]: There is no conflict here.
We agreed that the `Get*` functions are used to retrieve single entities, while `Query*` functions retrieve arrays.
If the function retrieve a single entity, let's call it `GetEntityBySomething` and use it consistently.

Comment (vbauzysvmware) [reply vcd/catalogitem.go]: Updated by new govcd changes

File: vcd/resource_vcd_catalog_media.go (hunk 9 above)
Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog_media.go]: changed

**Comment by vbauzysvmware** [comment]
Guys, updates are done. Should be now more cleaner.

File: vcd/datasource_vcd_catalog_media.go [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+				Computed:    true,
+				Description: "True if this media file is in a published catalog",
+			},
+			"creation_date": &schema.Schema{
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Creation date",
+			},
+			"size": &schema.Schema{
+				Type:        schema.TypeInt,
```
Comment (lvirbalas) [inline vcd/datasource_vcd_catalog_media.go]: Curious, why here we use `schema.TypeInt`, but in indendent disk "size" is `schema.TypeFloat`? Where's the caveat?

File: vcd/resource_vcd_catalog_media_test.go [hunk 13]
```diff
@@ -99,12 +125,12 @@ func testAccCheckVcdCatalogMediaExists(mediaName string, catalogItem *govcd.Cata
 			return fmt.Errorf("catalog %s does not exist (%s)", testSuiteCatalogName, err)
 		}
 
-		newCatalogItem, err := catalog.GetCatalogItemByName(catalogMediaRs.Primary.Attributes["name"], false)
+		newMedia, err := catalog.GetMediaByName(catalogMediaRs.Primary.Attributes["name"], false)
```
Comment (lvirbalas) [inline vcd/resource_vcd_catalog_media_test.go]: "new" and "old" are the evil twins of naming variables. Please use precise word which describes what this is, "mediaRecord" maybe (IIRC the govcd PR right)?

File: vcd/datasource_vcd_catalog_media.go (hunk 12 above)
Comment (vbauzysvmware) [reply vcd/datasource_vcd_catalog_media.go]: This isn't independent disk. type is ```	Size         int64            `xml:"size,attr,omitempty"` ```

File: vcd/resource_vcd_catalog_media_test.go (hunk 13 above)
Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog_media_test.go]: changed

File: vcd/datasource_vcd_catalog_media_test.go [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+    catalogMedia_metadata2 = "catalogMedia Metadata2"
+  }
+}
+
+data "vcd_catalog_media" "{{.NewCatalogMedia}}" {
+  org     = "{{.Org}}"
+  catalog = "{{.Catalog}}"
+  name    = "{{.CatalogMediaName}}"
+  depends_on = ["vcd_catalog_media.{{.CatalogMediaName}}"]
+}
```
Comment (dataclouder) [inline vcd/datasource_vcd_catalog_media_test.go]: I was initially confused by this setting, because the `NewCatalogMedia` suggests something being created, but the `data` means that we are using an existing resource. Then I noticed that the `name` element is the same as the resource above.

Since we use a media item for govcd tests, meaning that our servers already have one, why not checking whether any item exists and use it, like we do for vApps and networks?

File: vcd/datasource_vcd_catalog_media_test.go [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+		Providers:    testAccProviders,
+		CheckDestroy: catalogMediaDestroyed(testSuiteCatalogName, TestCatalogMediaDS),
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config:             configText,
+				ExpectNonEmptyPlan: true,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckVcdCatalogMediaExists("vcd_catalog_media."+TestAccVcdDataSourceMedia, &media),
+					resource.TestMatchOutput("owner_name", regexp.MustCompile(`^\w*$`)),
+					resource.TestMatchOutput("creation_date", regexp.MustCompile(`^(2019|2020)-`)),
```
Comment (dataclouder) [inline vcd/datasource_vcd_catalog_media_test.go]: This regexp is going to fail in 2021.
If that the expected date is in the format 'YYYY-MM-DD' I suggest using `^\d{4}-\d{2}-\d{2}`

File: vcd/datasource_vcd_catalog_media_test.go [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+		}
+		_, err = cat.GetMediaByName(mediaName, false)
+		if err == nil {
+			return fmt.Errorf("catalog media %s not deleted", mediaName)
+		}
+		return nil
+	}
+}
+
+func importStateIdByCatalogMedia(objectName string) resource.ImportStateIdFunc {
```
Comment (dataclouder) [inline vcd/datasource_vcd_catalog_media_test.go]: The function `importStateIdByCatalogItem` will return the same result. We can reuse it

File: vcd/resource_vcd_catalog_media.go [hunk 17]
```diff
@@ -132,39 +171,74 @@ func resourceVcdMediaCreate(d *schema.ResourceData, meta interface{}) error {
...
-	d.SetId(catalogName + ":" + mediaName)
-
 	log.Printf("[TRACE] Catalog media created: %#v", mediaName)
 
 	err = createOrUpdateMediaItemMetadata(d, meta)
 	if err != nil {
 		return fmt.Errorf("error adding media item metadata: %s", err)
 	}
 
+	//sets Id in findCatalogItem func - there isn't ID with media type
```
Comment (dataclouder) [inline vcd/resource_vcd_catalog_media.go]: Is this comment obsolete? There is an ID in media type

File: vcd/resource_vcd_catalog_media.go [hunk 18]
```diff
@@ -132,39 +171,74 @@ func resourceVcdMediaCreate(d *schema.ResourceData, meta interface{}) error {
...
+	if err != nil {
+		log.Printf("[DEBUG] Unable to find media: %s", err)
+		return err
+	}
+
+	d.SetId(media.Media.ID)
+
+	mediaRecord, err := catalog.QueryMedia(d.Get("name").(string))
+	if err != nil {
+		log.Printf("[DEBUG] Unable to query media: %s", err)
```
Comment (dataclouder) [inline vcd/resource_vcd_catalog_media.go]: This double request makes the code difficult to read.
I think we should change `GetMediaBy(Name|Id|Href)` to return a full media record, meaning that inside these functions we call `QueryMedia` and return a single structure containing everything we need.

File: vcd/resource_vcd_catalog_media.go [hunk 19]
```diff
@@ -211,18 +291,72 @@ func createOrUpdateMediaItemMetadata(d *schema.ResourceData, meta interface{}) e
...
+	media, err := catalog.GetMediaByName(mediaName, false)
+	if err != nil {
+		return nil, govcd.ErrorEntityNotFound
+	}
+
+	_ = d.Set("org", orgName)
+	_ = d.Set("catalog", catalogName)
+	_ = d.Set("name", mediaName)
+	_ = d.Set("description", media.Media.Description)
+	d.SetId(media.Media.ID)
```
Comment (dataclouder) [inline vcd/resource_vcd_catalog_media.go]: See my comment in the `Read` function. We are not setting several attributes because we lack the data.

File: website/docs/r/catalog_media.html.markdown [hunk 20]
```diff
@@ -45,3 +45,45 @@ The following arguments are supported:
...
+  org     = "my-org"
+  catalog = "my-vdc"
+  name    = "my-media"
+}
+```
+
+You can import such catalog media into terraform state using this command
+
+```
+terraform import vcd_catalog_media.tf-mymedia my-org.my-catalog-name.my-media-name
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply website/docs/r/catalog_media.html.markdown]: updated

File: website/docs/r/catalog_media.html.markdown [hunk 21]
```diff
@@ -45,3 +45,45 @@ The following arguments are supported:
...
+configuration. [More information.][docs-import]
+
+An existing media item can be [imported][docs-import] into this resource via supplying its path.
+The path for this resource is made of org-name.catalog-name.media-name
+For example, using this structure, representing a media item that was **not** created using Terraform:
+
+```hcl
+resource "vcd_catalog_media" "tf-mymedia" {
+  org     = "my-org"
+  catalog = "my-vdc"
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply website/docs/r/catalog_media.html.markdown]: yep, updated

File: vcd/datasource_vcd_catalog_media_test.go (hunk 15 above)
Comment (vbauzysvmware) [reply vcd/datasource_vcd_catalog_media_test.go]: Tx. updated

File: vcd/datasource_vcd_catalog_media_test.go (hunk 16 above)
Comment (vbauzysvmware) [reply vcd/datasource_vcd_catalog_media_test.go]: I liked to be separated-explicit, but updated

File: vcd/resource_vcd_catalog_media.go (hunk 17 above)
Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog_media.go]: yes, removed

File: vcd/resource_vcd_catalog_media.go (hunk 18 above)
Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog_media.go]: I would agree on having wrapper, cause not always you need `all the data` and thus create two request when you need one. And yet more votes needed for that, IMHO I am ok with this approach.

File: vcd/resource_vcd_catalog_media.go (hunk 19 above)
Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog_media.go]: Comment in above

File: vcd/datasource_vcd_catalog_media_test.go [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+		PreCheck:     func() { preRunChecks(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: catalogMediaDestroyed(testSuiteCatalogName, TestCatalogMediaDS),
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config:             configText,
+				ExpectNonEmptyPlan: true,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckVcdCatalogMediaExists("vcd_catalog_media."+TestAccVcdDataSourceMedia, &media),
+					resource.TestMatchOutput("owner_name", regexp.MustCompile(`^\w*$`)),
```
Comment (dataclouder) [inline vcd/datasource_vcd_catalog_media_test.go]: This regular expression means "match a string of  **zero** or more word characters". Thus it will match "abc", or "a_bc", but also the empty string "". However, it won't match "ab-c"

File: vcd/datasource_vcd_catalog_media_test.go [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+					resource.TestMatchOutput("creation_date", regexp.MustCompile(`^^\d{4}-\d{2}-\d{2}.*`)),
+					resource.TestCheckOutput("status", "RESOLVED"),
+					resource.TestMatchOutput("storage_profile_name", regexp.MustCompile(`(.|\s)*\S(.|\s)*`)),
```
Comment (dataclouder) [inline vcd/datasource_vcd_catalog_media_test.go]: This regular expression will easily match "a" or "aaa" or " a ".
`(.|\s)*` == **zero* or more occurrence of either any character or a space. 
The only mandatory character is `\S` which means "a non-space" character. Is this what we are trying to catch?

Comment (vbauzysvmware) [reply vcd/datasource_vcd_catalog_media_test.go]: My aim is non empty string, I can't guess/verify anything else

File: vcd/datasource_vcd_catalog_media_test.go (hunk 22 above)
Comment (vbauzysvmware) [reply vcd/datasource_vcd_catalog_media_test.go]: Do we now user input restrictions? or just check non empty string?

File: vcd/datasource_vcd_catalog_media_test.go (hunk 14 above)
Comment (vbauzysvmware) [reply vcd/datasource_vcd_catalog_media_test.go]: But this create very strange dependency - you have to run govcd tests before terraform test. Also you have option ignore configuring media in govcd and test will be ignored.

File: vcd/datasource_vcd_catalog_media_test.go (hunk 22 above)
Comment (vbauzysvmware) [reply vcd/datasource_vcd_catalog_media_test.go]: updated to `non-empty`

File: vcd/datasource_vcd_catalog_media_test.go (hunk 23 above)
Comment (vbauzysvmware) [reply vcd/datasource_vcd_catalog_media_test.go]: updated to `non-empty`

File: vcd/datasource_vcd_catalog_item_test.go [hunk 24]
```diff
@@ -96,8 +96,8 @@ func catalogItemDestroyed(catalog, itemName string) resource.TestCheckFunc {
 
 func importStateIdByCatalogItem(objectName string) resource.ImportStateIdFunc {
```
Comment (Didainius) [inline vcd/datasource_vcd_catalog_item_test.go]: I think existing more generic function `importStateIdOrgObject` could be used.

File: vcd/catalogitem.go [hunk 25]
```diff
@@ -55,15 +56,23 @@ func findCatalogItem(d *schema.ResourceData, vcdClient *VCDClient) (*govcd.Catal
...
+		log.Printf("[DEBUG] Unable to find catalog.")
 		return nil, fmt.Errorf("unable to find catalog: %s", err)
 	}
 
-	catalogItem, err := catalog.GetCatalogItemByName(d.Get("name").(string), false)
+	identifier := d.Id()
+
+	// Check if identifier is still in deprecated style `catalogName:mediaName`
+	// Required for backwards compatibility as identifier has been changed to vCD ID in 2.5.0
+	if identifier == "" || strings.Contains(identifier, ":") {
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply vcd/catalogitem.go]: Adjusted, as we agreed to use full Id's

File: vcd/resource_vcd_catalog_media_test.go [hunk 26]
```diff
@@ -54,12 +55,17 @@ func TestAccVcdCatalogMediaBasic(t *testing.T) {
 						"vcd_catalog_media."+TestAccVcdCatalogMedia, "metadata.mediaItem_metadata", "mediaItem Metadata"),
 					resource.TestCheckResourceAttr(
 						"vcd_catalog_media."+TestAccVcdCatalogMedia, "metadata.mediaItem_metadata2", "mediaItem Metadata2"),
+					resource.TestMatchOutput("owner_name", regexp.MustCompile(`^\S+`)),
+					resource.TestMatchOutput("creation_date", regexp.MustCompile(`^^\d{4}-\d{2}-\d{2}.*`)),
+					resource.TestCheckOutput("status", "RESOLVED"),
+					resource.TestMatchOutput("storage_profile_name", regexp.MustCompile(`^\S+`)),
+					testCheckMediaNonStringOutputs(),
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog_media_test.go]: no, it isn't - i tried

File: vcd/resource_vcd_catalog_media_test.go [hunk 27]
```diff
@@ -99,12 +125,12 @@ func testAccCheckVcdCatalogMediaExists(mediaName string, catalogItem *govcd.Cata
...
 
-		newCatalogItem, err := catalog.GetCatalogItemByName(catalogMediaRs.Primary.Attributes["name"], false)
+		foundMedia, err := catalog.GetMediaByName(catalogMediaRs.Primary.Attributes["name"], false)
 		if err != nil {
-			return fmt.Errorf("catalog media %s does not exist (%#v)", catalogMediaRs.Primary.ID, catalogItem.CatalogItem)
+			return fmt.Errorf("catalog media %s does not exist (%#v)", catalogMediaRs.Primary.ID, foundMedia.Media)
 		}
 
-		catalogItem = newCatalogItem
+		media = foundMedia
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog_media_test.go]: removed

File: vcd/resource_vcd_catalog_media_test.go [hunk 28]
```diff
@@ -153,7 +179,35 @@ const testAccCheckVcdCatalogMediaBasic = `
     mediaItem_metadata2 = "mediaItem Metadata2"
   }
 }
-`
+
+output "creation_date" {
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply vcd/resource_vcd_catalog_media_test.go]: could you rephrase

File: vcd/datasource_vcd_catalog_media_test.go [hunk 29]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+data "vcd_catalog_media" "{{.NewCatalogMedia}}" {
+  org     = "{{.Org}}"
+  catalog = "{{.Catalog}}"
+  name    = "{{.CatalogMediaName}}"
+  depends_on = ["vcd_catalog_media.{{.CatalogMediaName}}"]
+}
+
+output "size" {
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply vcd/datasource_vcd_catalog_media_test.go]: This for testing to see if variable is accessible and readed

File: vcd/datasource_vcd_catalog_media_test.go [hunk 30]
```diff
@@ -0,0 +1,147 @@
...
+			resource.TestStep{
+				Config:             configText,
+				ExpectNonEmptyPlan: true,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckVcdCatalogMediaExists("vcd_catalog_media."+TestAccVcdDataSourceMedia, &media),
+					resource.TestMatchOutput("owner_name", regexp.MustCompile(`^\S+`)),
+					resource.TestMatchOutput("creation_date", regexp.MustCompile(`^^\d{4}-\d{2}-\d{2}.*`)),
+					resource.TestCheckOutput("status", "RESOLVED"),
+					resource.TestMatchOutput("storage_profile_name", regexp.MustCompile(`^\S+`)),
+					testCheckMediaNonStringOutputs(),
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply vcd/datasource_vcd_catalog_media_test.go]: I needed cause Output doesn't work with other than strings

File: vcd/datasource_vcd_catalog_item_test.go (hunk 24 above)
Comment (vbauzysvmware) [reply vcd/datasource_vcd_catalog_item_test.go]: No it can't as it uses org:catalogname:objectname

**Comment by vbauzysvmware** [comment]
Code updated

File: vcd/datasource_vcd_catalog_media_test.go (hunk 30 above)
[the remark this answers is not in the document]
Comment (Didainius) [reply vcd/datasource_vcd_catalog_media_test.go]: Ok. I guess some of the output fields are not necessary here, but I am closing this discussion now and will raise it in future as a wider topic.

File: vcd/datasource_vcd_catalog_media_test.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (Didainius) [reply vcd/datasource_vcd_catalog_media_test.go]: Closing this discussion now and will raise it in future as a wider topic.

File: vcd/resource_vcd_catalog_media_test.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (Didainius) [reply vcd/resource_vcd_catalog_media_test.go]: Doesn't work for `output` fields. Closing it now.

File: vcd/resource_vcd_catalog_media_test.go (hunk 28 above)
[the remark this answers is not in the document]
Comment (Didainius) [reply vcd/resource_vcd_catalog_media_test.go]: I think we do not need "pass-through" `output` fields as this is introducing just more lines without any big use. Closing this discussion now and will raise it in future as a wider topic.

File: vcd/datasource_vcd_catalog_media_test.go (hunk 14 above)
Comment (dataclouder) [reply vcd/datasource_vcd_catalog_media_test.go]: What I mean is that we have already a media item created and pre-installed for the purpose of using it with govcd tests. The check I suggest should look for existing media and use it for data source testing.
With our pre-configured servers, the media item is always there.

File: website/docs/d/catalog_media.html.markdown [hunk 31]
```diff
@@ -0,0 +1,46 @@
...
+
+```hcl
+data "vcd_catalog_media" "existing-media" {
+  org     = "my-org"
+  catalog = "my-cat"
+  name    = "my-media"
+}
+
+output "media_size" {
+  value = data.vcd_catalog_media.size
```
Comment (Didainius) [inline website/docs/d/catalog_media.html.markdown]: Is this a valid data access? I think it is missing resource name `existing-media`. Also I do not think we need these `output` fields in examples. People can access them directly using `data.vcd_catalog_media.existing-media.size`

Comment (vbauzysvmware) [reply website/docs/d/catalog_media.html.markdown]: updated

File: vcd/datasource_vcd_catalog_media_test.go [hunk 32]
```diff
@@ -0,0 +1,144 @@
...
+	debugPrintf("#[DEBUG] CONFIGURATION: %s", configText)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { preRunChecks(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: catalogMediaDestroyed(testSuiteCatalogName, TestCatalogMediaDS),
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config:             configText,
+				ExpectNonEmptyPlan: true,
```
[the remark this answers is not in the document]
Comment (vbauzysvmware) [reply vcd/datasource_vcd_catalog_media_test.go]: updated

File: vcd/datasource_vcd_catalog_media_test.go (hunk 14 above)
Comment (vbauzysvmware) [reply vcd/datasource_vcd_catalog_media_test.go]: In current configuration we don't have media reference.
